lib/src/rules/no_duplicate_cases.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.no_duplicate_cases;
+
+import 'dart:collection';
+import 'package:analyzer/context/declared_variables.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/src/dart/constant/evaluation.dart';
+import 'package:analyzer/src/dart/constant/value.dart';
+import 'package:analyzer/src/generated/engine.dart';
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:linter/src/linter.dart';
+
+const desc = 'Do not use more the one case with same value';
+
+const details = r'''
+**DO NOT** use more the one case with same value. This can be
+of typo or changed value of constant.
+
+**GOOD:**
+
+```
+const int A = 1;
+switch (v) {
+  case A:
+  case 2:
+}
+```
+
+**BAD:**
+
+```
+const int A = 1;
+switch (v) {
+  case 1:
+  case 2:
+  case A:
+  case 2:
+}
+```
+''';
+
+class NoDuplicateCases extends LintRule {
+  NoDuplicateCases()
+      : super(
+            name: 'no_duplicate_cases',
+            description: desc,
+            details: details,
+            group: Group.errors);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  void visitSwitchStatement(SwitchStatement node) {
+    AnalysisContext context = node?.expression?.bestType?.element?.context;
+    if (context == null) {
+      return;
+    }
+    TypeProvider typeProvider = context.typeProvider;
+    TypeSystem typeSystem = context.typeSystem;
+    DeclaredVariables declaredVariables = context.declaredVariables;
+
+    Map<DartObjectImpl, List<SwitchCase>> clusters =
+        new HashMap<DartObjectImpl, List<SwitchCase>>(
+            equals: (DartObjectImpl key1, DartObjectImpl key2) {
+      DartObjectImpl equals = key1.equalEqual(typeProvider, key2);
+      return equals.isBool && equals.toBoolValue();
+    });
+
+    for (SwitchMember member in node.members) {
+      if (member is SwitchCase) {
+        Expression expression = member.expression;
+
+        DartObjectImpl result = expression.accept(new ConstantVisitor(
+            new ConstantEvaluationEngine(typeProvider, declaredVariables,
+                typeSystem: typeSystem),
+            rule.reporter));
+
+        if (result == null) {
+          continue;
+        }
+
+        List<SwitchCase> cluster = clusters[result];
+        if (cluster != null) {
+          cluster.add(member);
+        } else {
+          clusters[result] = <SwitchCase>[member];
+        }
+      }
+    }
+
+    for (List<SwitchCase> cluster in clusters.values) {
+      if (cluster.length > 1) {
+        cluster.forEach(rule.reportLint);
```

Comment:
I would probably choose to only generate a lint for every occurrence _after_ the first. That will minimize the amount of feedback without degrading the quality.

It would also allow for a slight simplification in the implementation: instead of keeping a list, you need only remember the first case that evaluated to a given value. If you find another case with the same value, you can report an error immediately without adding it to a list.